core/runtime/src/main/java/io/quarkus/runtime/logging/LoggingSetupRecorder.java

```diff
@@ -291,29 +287,22 @@ public static void initializeBuildTimeLogging(LogConfig config, LogBuildTimeConf
         InitialConfigurator.DELAYED_HANDLER.setBuildTimeHandlers(handlers.toArray(EmbeddedConfigurator.NO_HANDLERS));
     }
 
-    private static Level getLogLevel(String categoryName, CategoryConfig categoryConfig, Map<String, CategoryConfig> categories,
-            Level rootMinLevel) {
-        if (Objects.isNull(categoryConfig))
-            return rootMinLevel;
-
-        final InheritableLevel inheritableLevel = categoryConfig.level;
-        if (!inheritableLevel.isInherited())
-            return inheritableLevel.getLevel();
-
-        int lastDotIndex = categoryName.lastIndexOf('.');
-        if (lastDotIndex == -1)
-            return rootMinLevel;
-
-        String parent = categoryName.substring(0, lastDotIndex);
-        return getLogLevel(parent, categories.get(parent), categories, rootMinLevel);
-    }
-
-    private static CategoryBuildTimeConfig isSubsetOf(String categoryName, Map<String, CategoryBuildTimeConfig> categories) {
-        return categories.entrySet().stream()
-                .filter(buildCategoryEntry -> categoryName.startsWith(buildCategoryEntry.getKey()))
-                .map(Entry::getValue)
-                .findFirst()
-                .orElse(null);
+    public static <T> Level getLogLevel(String categoryName, Map<String, T> categories,
+            Function<T, InheritableLevel> levelExtractor, Level rootMinLevel) {
+        while (true) {
+            T categoryConfig = categories.get(categoryName);
+            if (categoryConfig != null) {
+                final InheritableLevel inheritableLevel = levelExtractor.apply(categoryConfig);
+                if (!inheritableLevel.isInherited()) {
+                    return inheritableLevel.getLevel();
+                }
+            }
+            final int lastDotIndex = categoryName.lastIndexOf('.');
+            if (lastDotIndex == -1) {
+                return rootMinLevel;
+            }
+            categoryName = categoryName.substring(0, lastDotIndex);
+        }
```

Comment:
I've rewritten the alorithm that goes up the category path to:
- use it for different `Config` objects (hence the `levelExtractor`)
- fix a gap in the original implementation that was falling back to the root (min-)level too early (e.g. if `com.foo.bar` has `.level` but no `min-level`, then we must continue with `com` in case `com.foo` is not configured)

/cc @galderz